pkg/autoscaler/multiscaler.go

```diff
@@ -213,6 +217,31 @@ func (m *MultiScaler) Inform(event string) bool {
 	}
 	return false
 }
+func (m *MultiScaler) updateRunner(ctx context.Context, runner *scalerRunner) {
+	runner.stopCh <- struct{}{}
```

Comment:
This is blocking since the channel is unbuffered. You probably want to close it instead and guard it from being closed multiple times.

Reply:
Initially I did try to close the channel but unit test detected race condition while accessing stopCh by multiple goroutines.

Reply:
> It's okay to block until the old interval has actually stopped (or rather got the signal to do so).

This is in the reconciliation hot path - thus adding a delay here delays the reconciliation loop. With the current defaults the delay has an upper bound of 2s

Reply:
If I take the path of closing `stopCh` channel, then there are further changes which will add delay as well.

```
close(runner.stopCh)

// Need mutex to update stopCh in runner.
runner.mux.Lock()

// need to create new channel for runner
runner.stopCh = make(chan struct{})

runner.mux.Unlock()
m.runScalerTicker(ctx, runner)
```

Reply:
It looks like the mutexes are already held here: https://github.com/knative/serving/blob/7a79af4fd89beac6e76bcbf426159a5581a9def1/pkg/autoscaler/multiscaler.go#L172

I'm not particular about the solution - just be aware of the delay

Reply:
Yes that lock is at one level higher. Updating channel reference would require a new lock at runner level.

Reply:
So the maximum delay added here is waiting for `tickScaler` to return, right? Will that add any significant delay?

Reply:
There should not be any significant delay added with `runScalerTicker` function because the function is using already initialized runner object and starting background process.